src/server/editorServices.ts

```diff
@@ -754,6 +756,106 @@ namespace ts.server {
             return false;
         }
 
+        acquireTypingForJs(project: Project) {
+            if (!this.host.getTsd) {
+                return;
+            }
+            const tsd = this.host.getTsd();
+            if (project === undefined || tsd === undefined) {
+                return;
+            }
+
+            const typingOptions: TypingOptions = project.projectOptions
+                ? project.projectOptions.typingOptions
+                : { enableAutoDiscovery: true };
+            const compilerOptions: CompilerOptions = project.projectOptions
+                ? project.projectOptions.compilerOptions
+                : undefined;
+            const projectRootPath: Path = project.isConfiguredProject() ? getDirectoryPath(<Path>project.projectFilename) : undefined;
+
+            // Todo: add support for local cache path
+            const cachePath = this.host.globalTypingCachePath;
+            const tsdJsonPath = toPath("tsd.json", cachePath, getCanonicalFileName);
+            const tsdJson = this.host.getTsdJson(tsdJsonPath);
+            const { cachedTypingPaths, newTypingNames } = JsTyping.discoverTypings(
+                sys,
+                project.getFileNames(),
+                projectRootPath,
+                /* safeListPath */ undefined,
+                tsdJson ? tsdJson.packageNameToTypingPath : undefined,
+                typingOptions,
+                compilerOptions
+            );
+
+            this.log("Cached typings include: " + cachedTypingPaths);
+            this.log("New typings to download: " + newTypingNames);
+
+            // Bail out when no actions are needed
+            if (cachedTypingPaths.length === 0 && newTypingNames.length === 0) {
+                return;
+            }
+
+            if (!sys.directoryExists(cachePath)) {
+                sys.createDirectory(cachePath);
+            }
+
+            // Inject the cached ones first
+            cachedTypingPaths.forEach(p => addTypingToProject(p, project));
+
+            if (newTypingNames && newTypingNames.length > 0) {
+                const typingsFolderPath = toPath("typings", this.host.globalTypingCachePath, getCanonicalFileName);
+                const query = new tsd.Query();
+                const api = tsd.getAPI(tsdJsonPath);
+                const options = new tsd.Options();
+                options.resolveDependencies = true;
+                options.overwriteFiles = true;
+                options.saveToConfig = true;
+
+                for (const newTypingName of newTypingNames) {
+                    query.addNamePattern(newTypingName);
+                }
+
+                let promise: any;

```

Comment:
Why are we typing this promise as `any`? That means you get no checking on any of the entire `.then(...` chain below.